CommandManager.cs

```diff
@@ -1,128 +1,192 @@
+        }
+        else
         {
-            foreach( var d in OnAnyCommandAsync.GetInvocationList( ).Cast<AsyncCommandDelegate>( ) )
+            methodInfo.Invoke( command.Instance, parsedArgs );
+        }
+    }
+
+    private static object[ ] ParseCommandArgs( IPlayer player, string[ ] args, MethodInfo methodInfo )
+    {
+        var parameters = methodInfo.GetParameters( );
+
+        // player should be first param at all times
+        var parsedArgs = new object[ parameters.Length ];
+        parsedArgs[ 0 ] = ( IPlayer ) player;
+
+        // Skip command name
+        args = args.Skip( 1 ).ToArray( );
+        var argCounter = 0;
+
+        // Loop all "OnCommand" method parameters, skip IPlayer(1st) param
+        for( int i = 1; i < parameters.Length; i++ )
+        {
+            // Failed to parse whatever the player sent, cancel here.
+            if( argCounter >= args.Length ) 
+                break;
+
+            // If its a string, we have some special parsing to do since it could be a sentence
+            if( parameters[ i ].ParameterType == typeof( string ) )
             {
-                await d.Invoke( player, args );
+                var leftoverArgs = args.Skip( argCounter );
+                
+                // If there is no element after this one, concat all remaining values into 1 string param
+                if( i + 1 >= parameters.Length )
+                {
+                    parsedArgs[ i ] = string.Join( " ", leftoverArgs );
+                    break;
+                }
+
+                parsedArgs[ i ] = ParseStringArg( parameters[ i + 1 ].ParameterType, leftoverArgs, ref argCounter );
+            }
+            // Easy, just a number or so
+            else
+            {
+                parsedArgs[ i ] = Convert.ChangeType( args[ argCounter ], parameters[ i ].ParameterType );
+                argCounter++;
             }
         }
-        
-        OnAnyCommand?.Invoke( player, args );
+
+        return parsedArgs;
     }
 
-    private void TriggerAnyKnownCommand( IPlayer player, string command, string[] args )
+    private static string ParseStringArg( Type nextParamType, IEnumerable<string> leftoverArgs, ref int argCounter )
     {
-        if( !_commandDelegates.TryGetValue( command, out var commandDelegate ) )
-            return;
+        var combinedArgs = new List<string>( );
+
+        foreach( var arg in leftoverArgs )
+        {
+            if( IsArgumentOfType( arg, nextParamType ) )
+                break;
+
+            argCounter++;
+            combinedArgs.Add( arg );
+        }
         
-        commandDelegate.Invoke( player, args );
+        return string.Join( " ", combinedArgs );
+    }
+
+    private static bool IsArgumentOfType( string arg, Type type )
+    {
+        try
+        {
+            _ = Convert.ChangeType( arg, type );
+            return true;
+        }
+        catch( Exception ex )
```

Comment:
We have a logger in here so i guess this would be a good place to logg the exception

Reply:
In this case nah, the Convert.ChangeType throws an exception if it can't convert it to the target type. 

This was, as far as I could find, the only quick way to determine whether a string value can be converted to the target type or not.